students/philip_behrend/lesson03/assignment/test_operations.py

```diff
@@ -0,0 +1,50 @@
+""" Tests for customer database module """
+from unittest import TestCase
+from customer_model import *
+from basic_operations import *
+from peewee import *
+import sqlite3, traceback
+
+class BasicOperationsTests(TestCase):
+    """ Tests for basic_operations file """
+
+    def test_add_customer(self):
+        add_customer(customer_id=9,
+          firstname='John',
+          lastname='Deere',
+          address='5555 Street',
+          phone_no='509-888-9999',
+          email='john.deere@gmail.com',
+          status=1,
+          credit_limit=4000)
+        cust = Customer.get_by_id(9)
+        self.assertEqual(cust.firstname, 'John')
+        self.assertEqual(cust.phone_no, '509-888-9999')
+
+    def test_add_invalid_customer(self):
+        with self.assertRaises(IntegrityError) as cm:
```

Comment:
This test not working is a side effect of the general exception handling in _basic_operations.py_. The exception is not being raised (which is what the test is expecting) because you _Except_ block catches all exceptions (including _IntegrityError_).

Reply:
Hi Luis, thanks for the pointer. I modified the try-except block to catch the IntegrityError specifically (given your statement above). However, the traceback states that there was a _separate_ exception in _new_cust = (Customer.create)(**kwargs)_. 
Do you have advice on how to troubleshoot why the try-except block isn't catching the seemingly obvious single exception (lastname = Null)?


def add_customer(**kwargs):
    """ Add customer function """
    try:
        with db.transaction():
            **new_cust = (Customer.create)(**kwargs)**
            new_cust.save()
            logger.info('Customer addition successful')
    except IntegrityError as e:
        logger.info('Error adding customer')
        logger.info(e)
        logger.info(traceback.format_exc())